docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+===========================================
+RFC 001: Storage Engines and Database Layer
+===========================================
+
+Changelog
+---------
+
+- 2021-04-19: Initial Draft (gist)
+- 2021-09-02: Migrated to RFC folder, with some updates  
+
+Abstract
+--------
+
+The aspect of Tendermint that's responsible for persistence and storage (often
+"the database" internally) represents a bottle neck in the architecture of the
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
+storage layer is handled generically, and most tests run with an in-memory
+framework, it's difficult to take advantage of any higher-level features of a
+database engine.
+
+Ideally, developers within Tendermint will be able to interact with persisted
+data via a single interface that can function, approximately like an object
+store, and this storage interface will be able to accommodate all existing
+persistence workloads (e.g. block storage, local peer management information
+like the "address book", crash-recovery log like the WAL.) In addition to
+providing a more ergonomic interface and new semantics, by selecting a single
+storage engine tendermint can use native durability and atomicity features of
+the storage engine and simplify its own implementations. 
+
+Data Access Patterns
+~~~~~~~~~~~~~~~~~~~~
+
+Tendermint's data access patterns have the following characteristics:
+
+- data size often exceeds memory.
+
+- data is rarely mutated after it's written for most data (e.g. blocks), but
+  small amounts of workding data is persisted by nodes and is frequently
+  mutated (e.g. peer information, validator information.)
+
+- read patterns can be quite random.
```

Comment:
Hmm. I don't disagree, but so far it seems like the read patterns are mostly fairly sequential: One might occasionally query for a single transaction (e.g., for debugging) but most of the Tendermint plumbing seems to orient around synchronizing logs between nodes. So the starting-point might be random(ish), but I think after that it's more or less a scan forward in time, right?

Reply:
I think logically you're correct, but I don't believe that the the pages are likely to be contiguous because the keys/leading bytes aren't monotonically increasing in the way that we'd want. 

I also suspect that for the purposes of consensus all of the data is (basically) is still in the application and we don't hit the db layer, and all other workloads (statesync, blocksync, query load from the application or RPC layers,) are really random.

Reply:
Yes, perhaps not strictly monotonic—that's a strong claim—but I meant in a sense that we could notionally expect the "next" datum in a sync process to have been recorded "near in time" to the previous one, since transactions are roughly ordered by when they were proposed. So we do have random access, but I think most accesses are (notionally) "within a cache line".

That said, I don't really understand how the app's internal state and the mempool and the persistent log all interact with each other, so I don't have a real difference to propose here.